File: Data Structures & Algorithms/design-word-search-data-structure/submission-3.py
class Node:
    def __init__(self):
        self.children = {}
        self.is_word = False

class WordDictionary:

    def __init__(self):
        self.root = Node()

    def addWord(self, word: str) -> None:
        curr = self.root
        for c in word:
            if c not in curr.children:
                curr.children[c] = Node()
            curr = curr.children[c]
        
        curr.is_word = True

    def search(self, word: str) -> bool:
        curr = self.root

        return self.recursive_search(curr, word)
    
    def recursive_search(self, node: Node, word: str) -> bool:
        if not node:
            return False
        if word == "":
            return node.is_word 
        
        for i, c in enumerate(word):
            if c == ".":
                for child in node.children.values():
                    if self.recursive_search(child, word[i+1:]): return True

                return False
            
            if c not in node.children:
                return False
            node = node.children[c]
        
        return node.is_word
        

File: Data Structures & Algorithms/implement-prefix-tree/submission-3.cpp
class PrefixTree {
private:
    struct Node {
        char val;
        std::unordered_map<char, Node*> children;
        bool is_word;

        Node() : val('\0'), is_word(false) {}
        Node(char val) : val(val), is_word(false) {}
    };
    Node* root;

    void clear(Node* node) {
        for (auto& child : node->children) {
            clear(child.second);
        }
        delete node;
    }

public:
    PrefixTree() {
        this->root = new Node();
    }
    ~PrefixTree() {
        clear(root);
    }
    
    void insert(string word) {
        Node* curr = this->root;

        for (const auto& c : word) {
            if (curr->children.find(c) == curr->children.end()) curr->children[c] = new Node(c);
            curr = curr->children[c];
        }
        curr->is_word = true;
    }
    
    bool search(string word) {
        Node* curr = this->root;

        for (const auto& c : word) {
            if (curr->children.find(c) == curr->children.end()) return false;
            curr = curr->children[c];
        }
        return curr->is_word;
    }
    
    bool startsWith(string prefix) {
        Node* curr = this->root;

        for (const auto& c : prefix) {
            if (curr->children.find(c) == curr->children.end()) return false;
            curr = curr->children[c];
        }
        return true;
    }
};
